### DISCUSSION TITLE
Can Mojo be used to create C/C++ static and dynamic libraries?

### DISCUSSION THREAD

**Discussion by zhwa** [modular/modular#303] [Ideas]
In the [roadmap](https://docs.modular.com/mojo/roadmap.html#cc-interop), it is mentioned that Mojo will be able to leverage C/C++ modules. I wonder if in the future, a Mojo program can be compiled into static and dynamic libraries that can be consumed by C/C++ projects?

**Comment by mojodojodev** [comment]
Yes absolutely, see here: https://mojodojo.dev/mojo_team_answers.html#producing-binaries

**Reply by slagernate** [reply]
A simple example would be most appreciated :D

**Comment by lattner** [comment]
Yep!

**Comment by SeeRich** [comment]
Is there any timeframe on this? I would love to be able to call mojo functions through another languages C-FFI (e.g. Rust).

**Reply by gabrieldemarmiesse** [reply]
I think it's already available, from the docs:
> Global variables can now be exported in Mojo compiled archives, using the `@export` decorator. Exported global variables are public symbols in compiled archives and use the variable name as its linkage name, by default. A custom linkage name can be specified with `@export("new_name")`. This does not affect variable names in Mojo code.

**Reply by SeeRich** [reply]
Good find, this may be what I’m looking for! However, I don’t see any other references to “compiled archives” or any example/documentation for using the `@export` decorator. Ultimately, I would need to compile to a shared library so the mojo code could be consumed by another executable.

**Comment by nevakrien** [comment]
So I looked into it a while ago I got the answer not yet I would like to hear an update

**Comment by hogepodge** [comment]
You're right that the extern keyword is there, but it's not something that we've documented beyond the release notes.

The reason being is that the feature isn't fully baked yet, and we don't want to encourage usage that will likely change in the future and whose usage would feel like a hack.

**Reply by slagernate** [reply]
Thanks for the update!
